### PR TITLE
8343068: C2: CastX2P Ideal transformation not always applied

### PR DESCRIPTION
**Pull request by rwestrel**
The transformation:

```
(CastX2P (AddL base i)) -> (AddP (CastX2P base) i) 
```

when i fits in an int is not always applied: when the type of `i` is
narrowed so it fits in an int, the `CastX2P` is not enqueued for
igvn. This can get in the way of vectorization as shown by test case
`test2`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343068](https://bugs.openjdk.org/browse/JDK-8343068): C2: CastX2P Ideal transformation not always applied (**Bug** - P5)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21714/head:pull/21714` \
`$ git checkout pull/21714`

Update a local copy of the PR: \
`$ git checkout pull/21714` \
`$ git pull https://git.openjdk.org/jdk.git pull/21714/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21714`

View PR using the GUI difftool: \
`$ git pr show -t 21714`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21714.diff">https://git.openjdk.org/jdk/pull/21714.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21714#issuecomment-2437931952)
</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/c2/TestCastX2NotProcessedIGVN.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
+package compiler.c2;/*
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/c2/TestCastX2NotProcessedIGVN.java:1]: Please start the copyright header in a new line.

Comment (rwestrel) [reply test/hotspot/jtreg/compiler/c2/TestCastX2NotProcessedIGVN.java]: Right! fixed in new commit.

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by TobiHartmann — COMMENTED** [review]
Shouldn't this be caught by `VerifyIterativeGVN` after [JDK-8298952 ](https://bugs.openjdk.org/browse/JDK-8298952)?

**Review by TobiHartmann — APPROVED** [review]

**Comment by TobiHartmann** [comment]
Ah right, I missed that.

**Pull request closed without merging**